tempesta_fw/tfw_tfwcfg.c

```diff
@@ -0,0 +1,676 @@
+			loc->arg = NULL;
+		}
+		if (loc->cam) {
+			kfree(loc->cam);
+			loc->cam = NULL;
+		}
+	}
+	for (i = 0; i < tfwcfg_camatch_sz; ++i) {
+		TfwCfgCacheMatch *cam = &tfwcfg_camatch[i];
+		if (cam->arg) {
+			kfree(cam->arg);
+			cam->arg = NULL;
+		}
+	}
+}
+
+static void
+tfwcfg_cleanup_locache(TfwCfgSpec *cs)
+{
+	__tfwcfg_cleanup_locache();
+}
+
+static void
+tfwcfg_location_print_one(int idx, TfwCfgLocation *loc)
+{
+	int i;
+
+	printk(KERN_ERR "%s: [%d]: location op [%d] arg [%s] len [%d]"
+		" cam [%p] cam_sz [%d] cam_max [%d]\n",
+		__func__, idx, loc->op, loc->arg, loc->len,
+		loc->cam, loc->cam_sz, loc->cam_max);
+
+	if (!loc->cam)
+		return;
+
+	for (i = 0; i < loc->cam_sz; ++i) {
+		TfwCfgCacheMatch *cam = loc->cam[i];
+		printk(KERN_ERR "    %s: [%d]: cache match stmt [%d]"
+			" op [%d] arg [%s] len [%d]\n",
+			__func__, i, cam->stmt, cam->op,
+			cam->arg, cam->len);
+	}
+}
+static void
+tfwcfg_location_print_all(void)
+{
+	int i;
+
+	printk(KERN_ERR "%s: tfwcfg_location_sz [%d] tfwcfg_location_max [%d]\n",
+		__func__, tfwcfg_location_sz, tfwcfg_location_max);
+
+	for (i = 0; i < tfwcfg_location_sz; ++i) {
+		TfwCfgLocation *loc = &tfwcfg_location[i];
+		tfwcfg_location_print_one(i, loc);
+	}
+
+	printk(KERN_ERR "%s: default location\n", __func__);
+	tfwcfg_location_print_one(0, &tfwcfg_location_dflt);
+}
+
+static int
+tfwcfg_start(void)
+{
+tfwcfg_location_print_all();
+	return 0;
+}
+
+static void
+tfwcfg_stop(void)
+{
+	__tfwcfg_cleanup_locache();
+}
+
+static TfwCfgSpec tfwcfg_location_specs[] = {
+        {
+		"cache_bypass", NULL,
+		tfwcfg_handle_in_cache_bypass,
+		.allow_none = true,
+		.allow_repeat = true,
+		.cleanup = tfwcfg_cleanup_locache
+        },
+        {
+		"cache_fulfill", NULL,
+		tfwcfg_handle_in_cache_fulfill,
+		.allow_none = true,
+		.allow_repeat = true,
+		.cleanup = tfwcfg_cleanup_locache
+        },
+        {}
+};
+
+static TfwCfgSpec tfwcfg_specs[] = {
+	{
+		"cache_bypass", NULL,
+		tfwcfg_handle_out_cache_bypass,
+		.allow_none = true,
+		.allow_repeat = true,
+		.cleanup = tfwcfg_cleanup_locache
+	},
+        {
+		"cache_fulfill", NULL,
+		tfwcfg_handle_out_cache_fulfill,
+		.allow_none = true,
+		.allow_repeat = true,
+		.cleanup = tfwcfg_cleanup_locache
+        },
+	{
+		"location", NULL,
+		tfw_cfg_handle_children,
+		tfwcfg_location_specs,
+		&(TfwCfgSpecChild) {
+			.begin_hook = tfwcfg_begin_location,
+			.finish_hook = tfwcfg_finish_location
+		},
+		.allow_none = true,
+		.allow_repeat = true,
+		/* .cleanup function in a section with
+		   children causes a BUG_ON in cfg.c. */
+	},
+	{},
+};
+
+TfwCfgMod tfw_tfwcfg_cfg_mod = {
+	.name	= "tfwcfg",
+	.start	= tfwcfg_start,
+	.stop	= tfwcfg_stop,
+	.specs	= tfwcfg_specs,
+};
+
+int
+tfw_tfwcfg_init(void)
+{
+	int size;
+
+	BUG_ON(tfwcfg_camatch);
+	BUG_ON(tfwcfg_location);
+
+	/* Array of location directives. */
+	size = sizeof(TfwCfgLocation) * TFW_LOCATION_ARRAY_SZ;
+	tfwcfg_location = kzalloc(size, GFP_KERNEL);
```

Comment:
Why to allocate the array dynamically if we can allocate it statically at compile time? Usually compiler do better optimization work if static data is used, so always use static data if possible.